File: Arrays/169. Majority Element/169. Majority Element.py
# Boyer-Moore Majority Vote Algorithm
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        current_max = None
        count = 0
        for n in nums:
            if n == current_max:
                count += 1
            elif count == 0:
                current_max = n
                count = 1
            else:
                count -= 1
        return current_max

# Time: O(n)
# Space: O(1)

File: Arrays/169. Majority Element/169. Majority Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */

// Boyer-Moore Majority Vote Algorithm
var majorityElement = function (nums) {
  let currentMax = -Infinity;
  let count = 0;

  for (let n of nums) {
    if (n === currentMax) {
      count++;
    } else if (count === 0) {
      currentMax = n;
      count = 1;
    } else {
      count--;
    }
  }

  return currentMax;
};

// Time: O(n)
// Space: O(1)
